src/Microsoft.AspNetCore.Http/StreamPipeWriter.cs

```diff
@@ -0,0 +1,313 @@
+            }
+        }
+
+        /// <inheritdoc />
+        public override void OnReaderCompleted(Action<Exception, object> callback, object state)
+        {
+            if (callback == null)
+            {
+                throw new ArgumentNullException(nameof(callback));
+            }
+
+            // Note this currently doesn't do anything.
+            // Implementing completions/callbacks would require creating a PipeReaderAdapter.
+        }
+
+        /// <inheritdoc />
+        public override ValueTask<FlushResult> WriteAsync(ReadOnlyMemory<byte> source, CancellationToken cancellationToken = default)
+        {
+            return base.WriteAsync(source, cancellationToken);
+        }
+
+        /// <inheritdoc />
+        public override ValueTask<FlushResult> FlushAsync(CancellationToken cancellationToken = default)
+        {
+            return FlushAsyncInternal(cancellationToken);
+        }
+
+        private void Cancel()
+        {
+            InternalTokenSource.Cancel();
+        }
+
+        private async ValueTask<FlushResult> FlushAsyncInternal(CancellationToken cancellationToken = default)
+        {
+            if (_bytesWritten == 0)
+            {
+                return new FlushResult(isCanceled: false, IsCompletedOrThrow());
+            }
+
+            try
+            {
+                CancellationTokenRegistration reg;
+                if (cancellationToken.CanBeCanceled)
+                {
+                    reg = cancellationToken.Register(state => ((StreamPipeWriter)state).Cancel(), this);
+                    using (reg)
+                    {
+                        return await WriteAndFlushAsync();
+
+                    }
+                }
+                else
+                {
+                    return await WriteAndFlushAsync();
+                }
+            }
+            finally
+            {
+                _bytesWritten = 0;
+                _position = 0;
+            }
+        }
+
+        private async ValueTask<FlushResult> WriteAndFlushAsync()
+        {
+            try
+            {
+                // Write all completed segments and whatever remains in the current segment
+                // and flush the result.
+                if (_completedSegments != null && _completedSegments.Count > 0)
+                {
+                    var count = _completedSegments.Count;
+                    for (var i = 0; i < count; i++)
+                    {
+                        var segment = _completedSegments[i];
+#if NETCOREAPP2_2
+                        await _writingStream.WriteAsync(segment.Buffer.Slice(0, segment.Length), InternalTokenSource.Token);
+#elif NETSTANDARD2_0
+                        await _writingStream.WriteAsync(segment.Buffer.ToArray(), 0, segment.Length, InternalTokenSource.Token);
+#else
+#error Target frameworks need to be updated.
+#endif
+                        segment.Return();
```

Comment:
You need to test more cancellation scenarios. Current logic would do crazy things if you continue writing after CancelPendingFlush (you always reset _bytesWritten but clear segments only after writing them all, this might cause double writes/double returns/etc).

Reply:
Still adding a few more tests.